### PR TITLE
Cortex-R MPU support

### PR DESCRIPTION
**Pull request by bbolen**
The first three commits in this PR just move the Cortex-M mpu code around to make room for the Cortex-R support.  Let me know if it makes more sense for those 3 commits to be a separate pull request.  The rest of the commits add Cortex-R MPU and userspace support.  This was tested on a nucleo f746zg platform.

### CODE REVIEW

File: arch/arm/core/aarch32/mpu/arm_mpu.c
Comment (andrewboie) [inline arch/arm/core/aarch32/mpu/arm_mpu.c:61]: style: can you encapsulate these snippets of ASM into their own (maybe inline) functions? MISRA-C I believe requires this.

I think it would be good to have some accessor methods for these `mrc` operations

Comment (andrewboie) [reply arch/arm/core/aarch32/mpu/arm_mpu.c:61]: It looks like you did this in arm_mpu_internal.h so maybe, more like that

File: arch/arm/core/aarch32/userspace.S
Comment (andrewboie) [inline arch/arm/core/aarch32/userspace.S:464]: this is all good, it's undefined behavior to invoke a syscall trap from supervisor mode on any arch, and syscalls themselves need to be preemptible

**Review by tyrantlizard77 — COMMENTED** [review]
still looking this over but overall looking good. a few style nits

**Comment by bbolen** [comment]
@andrewboie Have you had a chance to look this over any more?

**Comment by carlescufi** [comment]
@sandeepbrcm @Abhishek-brcm @carlocaione is there a chance you can help out with the review?

**Comment by andrewboie** [comment]
I will take a further look this week.

File: include/arch/arm/aarch32/arch.h
Comment (andrewboie) [inline include/arch/arm/aarch32/arch.h:186]: I think it's necessary for you to squish this into previous commit so that the tree is still bisectable.

File: arch/arm/core/aarch32/thread.c
Comment (andrewboie) [inline arch/arm/core/aarch32/thread.c:278]: Would `if (arch_is_user_context())` work for all cases, not just Cortex-R?

File: arch/arm/core/aarch32/mpu/arm_mpu_v7_internal.h
Comment (andrewboie) [inline arch/arm/core/aarch32/mpu/arm_mpu_v7_internal.h:175]: yeah this change looks necessary.
I wonder if this issue this fixes is reproducible in the NXP or ARMv8 version of this MPU code.

**Review by tyrantlizard77 — COMMENTED** [review]
Overall looks very nice. A few comments.

**Comment by andrewboie** [comment]
@ioannisg would value your review on this PR.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by amergnat — COMMENTED** [review]
Nice job.
I admit to be a bit lost for some AMR mechanism, for example:
- Decrement nested count wasn't managed /supported before ?
- Swap change with r3 to un-do stack pointer alignment to double-word boundary
...

Then I can't approve, but consider my +1. ;)

File: arch/arm/core/aarch32/mpu/arm_mpu.c
Comment (ruuddw) [inline arch/arm/core/aarch32/mpu/arm_mpu.c:89]: Below get_region_attr_from_k_mem_partition_info seems to abstract from the MPU region attribute details (e.g. v7 or v8), any specific reason to handle cortex R differently with an #ifdef here instead?

File: include/arch/arm/aarch32/arch.h
Comment (bbolen) [reply include/arch/arm/aarch32/arch.h:186]: I left a compatibility layer in include/arch/arm/aarch32/cortex_m/mpu that includes the header files from the new location.  I went back and did builds after the 1st commit, after the second commit, and then after the 3rd to make sure they still built.  Unless I missed something, it should still be bisectable.

Comment (andrewboie) [reply include/arch/arm/aarch32/arch.h:186]: oh, missed that. thanks!

File: arch/arm/core/aarch32/thread.c
Comment (bbolen) [reply arch/arm/core/aarch32/thread.c]: arch_is_user_context() doesn't work for cortex-M.  I tried it and it broke the mem_protect/userspace test on the mps2_an385 board.  arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode.  Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level.  This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.

File: arch/arm/core/aarch32/mpu/arm_mpu.c
Comment (bbolen) [reply arch/arm/core/aarch32/mpu/arm_mpu.c]: no reason. I updated it so that attr is abstracted under get_region_attr_from_k_mem_partition_info

**Comment by andrewboie** [comment]
> arch_is_user_context() doesn't work for cortex-M. I tried it and it broke the mem_protect/userspace test on the mps2_an385 board. arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode. Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level. This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.

Ugh. This has the stench of something that is not well defined.

What is the difference between arch_is_user_context() and arch_is_unprivileged_state()??

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
NACK on adding `arch_is_unprivileged_state`. This either needs to be handled entirely within the arch/arm code or we have a bug here.

**Comment by bbolen** [comment]
> > arch_is_user_context() doesn't work for cortex-M. I tried it and it broke the mem_protect/userspace test on the mps2_an385 board. arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode. Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level. This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.
> 
> Ugh. This has the stench of something that is not well defined. Because this function is literally supposed to report "are we in user mode" with nothing else attached
> 
> What is the difference between arch_is_user_context() and arch_is_unprivileged_state()? what does this extra check of IPSR for?
> 
> this really seems like a bug to be honest

https://github.com/zephyrproject-rtos/zephyr/pull/17438 gives background of this arch_is_user_context() change.  Cortex-M arch_is_user_context returns true if the nPRIV bit is set AND the IPSR register is in thread mode (not in any other exception mode, i.e. fault, systick, irq).  The checks in fatal.c and thread.c only care that the nPRIV bit is set.  I thought the new function was the direction #17438 was headed although I understand your trepidation.  I'm not sure which direction to go from here though.  Maybe @ioannisg has some thoughts as well?

**Comment by andrewboie** [comment]
> #17438 gives background of this arch_is_user_context() change. Cortex-M arch_is_user_context returns true if the nPRIV bit is set AND the IPSR register is in thread mode (not in any other exception mode, i.e. fault, systick, irq). The checks in fatal.c and thread.c only care that the nPRIV bit is set. I thought the new function was the direction #17438 was headed although I understand your trepidation. I'm not sure which direction to go from here though. Maybe @ioannisg has some thoughts as well?

I'm still confused.
`arch_is_user_context()` is supposed to return false if you are in an exception or interrupt, because you aren't in user mode, you're in supervisor mode handling an exception.

If we are in supervisor mode handling a syscall/exception/irq and want to know if the interrupted context was in user mode we can just check if `(_current->base.user_options & K_USER)` anyway if we need to.

If this situation needs special handling for ARM purposes you need to deal with this stuff entirely within `arch/arm`. Hard NACK on this new arch interface. Might as well revert your PR back to before you added it.

**Comment by bbolen** [comment]
> Might as well revert your PR back to before you added it.
I reverted this back to the previous state.

I did noticed that z_check_thread_stack_fail is currently unused for Cortex-R.  I could define that out for Cortex-R and side step this issue for now.  It could be revisited when better fault decoding for Cortex-R is looked at.

**Comment by bbolen** [comment]
@andrewboie What do I need to do to get this PR moving along again?  I have floating point support for cortex-r that I would like to submit soon, but it depends on this first.

**Comment by bbolen** [comment]
@ioannisg I haven't seen Andrew around here in a while.  Do you know how I can get some traction again on this PR?  Do I need to get the attention of any particular person?  Thanks.

**Comment by carlocaione** [comment]
> @ioannisg I haven't seen Andrew around here in a while. Do you know how I can get some traction again on this PR? Do I need to get the attention of any particular person? Thanks.

maybe @dcpleung ?

**Comment by dcpleung** [comment]
LGTM. But there are lots of ARM assembly that I am not familiar with. It is better to get the ARM maintainers to go through this.

**Comment by carlocaione** [comment]
Out of curiosity, how difficult is to reuse this code for the Cortex-R82 given that we have #33905 ? Tagging @povergoing

**Comment by povergoing** [comment]
> Out of curiosity, how difficult is to reuse this code for the Cortex-R82 given that we have #33905 ? Tagging @povergoing

Actually, I had thought to reuse aarch32 MPU as the first plan. But the cortex-R82's MPU registers are 64bit, many `#define` interface need to be refined. And since aarch32 MPU has already been too complicated, and not too much code is reusable, so I give it up on the first step. 
It's a big work, but I could investigate it further. If it's possible, maybe we could discuss it.
P.S. or maybe it's even possible to combine MPU and MMU together.

**Comment by carlocaione** [comment]
> Actually, I had thought to reuse aarch32 MPU as the first plan. But the cortex-R82's MPU registers are 64bit, many `#define` interface need to be refined. And since aarch32 MPU has already been too complicated, and not too much code is reusable, so I give it up on the first step.

Ok, good to know.

> It's a big work, but I could investigate it further. If it's possible, maybe we could discuss it.

As you prefer but it's definitely not high priority. Thanks.

**Comment by bbolen** [comment]
@carlocaione @ioannisg @stephanosio @galak @MaureenHelm I'm tagging the people marked as owners in arch/arm so hopefully I didn't get the wrong people.  Can I get a review on this?  I have floating point support for 32 bit Cortex A/R that I think would be beneficial to zephyr, but it depends on this changeset and I cannot submit it.

**Review by julien-massot — APPROVED** [review]
Applied on R7 based platform and I'm able to run hello_world_user successfully.

File: include/arch/arm/aarch32/cortex_a_r/mpu.h
Comment (bbolen) [reply include/arch/arm/aarch32/cortex_a_r/mpu.h:5]: fixed

File: include/arch/arm/aarch32/cortex_a_r/mpu.h
Comment (julien-massot) [reply include/arch/arm/aarch32/cortex_a_r/mpu.h]: thanks

**Comment by carlescufi** [comment]
@npitre I'd be very grateful if you could take a quick look at this.

**Comment by ioannisg** [comment]
Will review this so it gets into 2.6, hopefully
@bbolen can you rebase?

**Comment by bbolen** [comment]
> Will review this so it gets into 2.6, hopefully
> @bbolen can you rebase?

yes, done. thanks.  note that the compliance test is failing on the check-warns test.  It is complaining about the first 3 commits which are just moving code.  I didn't actually add code to cause those warnings to appear.

**Comment by npitre** [comment]
I'm not very familiar with Cortex-R but here's a few random things I noticed:

```
+    /* validate syscall limit, only set priv mode if valid */
+    ldr ip, =K_SYSCALL_LIMIT
+    cmp r6, ip
+    blt valid_syscall_id
```

Here `lt` is a signed comparison flag. This means any negative number will be
considered valid.

...

```
+#if defined(CONFIG_CPU_CORTEX_R)
+               if (arch_is_user_context() == false)
+#else
+               if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == 0U)
+#endif
```

Conditional if statements like the above are confusing. It would be prettier if
the second case was abstracted into a `arch_is_user_context()` of its own too.

...

```
                if (is_user_accessible_region(r_index, write)) {
-                       return 0;
+                       r_index = 0;
                } else {
-                       return -EPERM;
+                       r_index = -EPERM;
                }
+               arch_irq_unlock(key);
+               return r_index;
```

That's an abuse of the `r_index` variable. You are not returning any index
here. Please use another variable that clearly identify its content. They are free and the compiler will
most likely generate the same code anyway.

...

```
-    pop {r0, r1}
+    pop {r0, r1}  /* saved just after valid_syscall_id */
+    pop {r2, r3}  /* saved just after z_arm_svc */
```

Can't you combine those i.e. `pop {r0 - r3}`?
A comment should still explain each pair.

...

```
+zephyr_library_include_directories(.)
```

What's the point of this? (no pun intended)

...

```
+    /*
+     * We need to store the spsr_svc onto the user stack before going off to
+     * the system call dispatcher.  This is needed since the system call may
+     * call _swap() which will invoke another SVC, overwriting this register.
+     */
```

Is this safe? What happens if the user stack pointer is corrupted?
Or worse: if it is pointing at some kernel memory that controls this user's
privileges?

About those commits:

- arch: arm: userspace: Use rfeia to return from syscall

- arch: arm: userspace: Reload r2 and r3 returning from system calls

- arch: arm: userspace: Decrement nested count before calling syscall

They appear to be fixes for changes that were added by earlier commits in the
series. If so they should be combined with those commits directly before being
merged upstream.

**Comment by bbolen** [comment]
Thank you for reviewing this.  I have fixed up all but one of your comments.

> +    /*
> +     * We need to store the spsr_svc onto the user stack before going off to
> +     * the system call dispatcher.  This is needed since the system call may
> +     * call _swap() which will invoke another SVC, overwriting this register.
> +     */
> ```
> Is this safe? What happens if the user stack pointer is corrupted?
> Or worse: if it is pointing at some kernel memory that controls this user's
> privileges?

This does appear to be an issue.  I've engaged the developer here that added the MPU support and we're trying to come up with a solution.

**Comment by ioannisg** [comment]
> The first three commits in this PR just move the Cortex-M mpu code around to make room for the Cortex-R support. Let me know if it makes more sense for those 3 commits to be a separate pull request

If this is not a problem, please do that @bbolen  and I'll take a look immediately. It's cleaner this way.

**Comment by ioannisg** [comment]
> This does appear to be an issue. I've engaged the developer here that added the MPU support and we're trying to come up with a solution.

Do we, or could we have a test that exposes this scenario, so we can verify that will be addressing this in the right way?

**Comment by ioannisg** [comment]
Needs rebase now @bbolen

File: arch/arm/core/aarch32/fatal.c
Comment (ioannisg) [inline arch/arm/core/aarch32/fatal.c:86]: can we specify in the name that we are talking about thread mode being nPRIV?
Cause, otherwise we have the arch_is_user_context API, but that one checks the handler mode as well..

Comment (ioannisg) [reply arch/arm/core/aarch32/fatal.c:86]: E.g. aarch32_thread_is_user_mode(), or sthg similar

File: include/arch/arm/aarch32/syscall.h
Comment (ioannisg) [inline include/arch/arm/aarch32/syscall.h:198]: This is fine (see, though my comment about the function name)

File: include/arch/arm/aarch32/syscall.h
Comment (ioannisg) [inline include/arch/arm/aarch32/syscall.h:181]: Can you use the Cortex-M CMSIS function, __get_CONTROL() here?

File: arch/arm/core/aarch32/fatal.c
Comment (bbolen) [reply arch/arm/core/aarch32/fatal.c]: fixed

File: include/arch/arm/aarch32/syscall.h
Comment (bbolen) [reply include/arch/arm/aarch32/syscall.h]: fixed

File: include/arch/arm/aarch32/syscall.h
Comment (bbolen) [reply include/arch/arm/aarch32/syscall.h]: fixed

**Comment by bbolen** [comment]
> +    /*
> +     * We need to store the spsr_svc onto the user stack before going off to
> +     * the system call dispatcher.  This is needed since the system call may
> +     * call _swap() which will invoke another SVC, overwriting this register.
> +     */
> ```
> 
> Is this safe? What happens if the user stack pointer is corrupted?
> Or worse: if it is pointing at some kernel memory that controls this user's
> privileges?
> 
I waited until v2.6 came out to get this out again.  All of your concerns should be addressed now.

File: tests/kernel/mem_protect/syscalls/src/main.c
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:322]: @bbolen this looks more like a unit test for the cortex-r architecture; i wonder if it could reside as a standalone unit test under tests/arch/arm

Comment (bbolen) [reply tests/kernel/mem_protect/syscalls/src/main.c]: agreed.  i've moved it and updated the PR

File: arch/arm/core/aarch32/mpu/CMakeLists.txt
Comment (ioannisg) [inline arch/arm/core/aarch32/mpu/CMakeLists.txt:11]: This could be an else if, as well. Potentially adding an else with an error message, also.

File: arch/arm/core/aarch32/fatal.c
Comment (ioannisg) [inline arch/arm/core/aarch32/fatal.c:86]: This is refactoring work. @bbolen can you quickly send this in a separate PR that could be reviewed and merged asap?

File: arch/arm/core/aarch32/mpu/arm_mpu.c
Comment (ioannisg) [inline arch/arm/core/aarch32/mpu/arm_mpu.c:45]: This is also refactoring. I am seeing this is now moved to the internal headers, as well. Can we move this to the refactoring PR?

File: arch/arm/core/aarch32/thread.c
Comment (ioannisg) [inline arch/arm/core/aarch32/thread.c:402]: this one as well.

File: include/arch/arm/aarch32/thread.h
Comment (ioannisg) [inline include/arch/arm/aarch32/thread.h:124]: Might have missed that, but it looked to me these are only used in Cortex-R, so could we #ifdef them out from cortex-M, here?

**Review by ioannisg — COMMENTED** [review]
@bbolen I did a quick pass, and noticed that there's still some ARM code refactoring done here.
I'd appreciate if we could shrink this diff, taking out the refactoring done, in another PR that could be reviewed faster and merged (hopefully this week). 

I added some comments inline, to help this out.

**Comment by ioannisg** [comment]
@stephanosio I'd appreciate your review here :)

File: arch/arm/core/aarch32/mpu/arm_mpu.c
Comment (bbolen) [reply arch/arm/core/aarch32/mpu/arm_mpu.c:45]: Any chance we can leave this one alone?  Pulling this out into a separate commit is going to be pretty tricky.

**Comment by bbolen** [comment]
@ioannisg @stephanosio You guys have any time to look at this?  I don't want to miss the feature window closure this time and it's coming up quickly.

**Comment by bbolen** [comment]
> @bbolen - are @ioannisg 's previous comments resolved?

@cfriedt  Yes.  I'll can mark them as resolved.  I've been confused on whether the submitter or reviewer should do that.

**Comment by cfriedt** [comment]
> > @bbolen - are @ioannisg 's previous comments resolved?
> 
> @cfriedt  Yes.  I'll can mark them as resolved.  I've been confused on whether the submitter or reviewer should do that.

Agreed. It is not very clear whose responsibility that is - I think for now either is fine. Sometimes it resolve myself after examining code / comments just to keep things moving.

Thanks!

**Comment by ruuddw** [comment]
I'm really not a ARM expert, so it's not up to me to approved this. It looks fine to me with all the feedback addressed by bbolen. From what I see, Andrew's NACK on arch_is_unprivileged_state has been resolved: his -1 can be overridden or ignored I think.
Why the "user_string_nlen" needs special handling for ARM isn't really clear to me - but it is not something introduced in this PR, just extended/improved. 

There is a fix for a missing critical section in mpu_buffer_validate(). @IRISZZW, @evgeniy-paltsev : arc v2 mpu has similar code and no lock either, can you please check and see if a lock should be added? The arc v4 mpu does have the lock, but no loop so maybe not required there? Maybe something to look into when extending the ARC v2/v4 MPU support with v3/v6.

**Comment by cfriedt** [comment]
@ioannisg - it seems as though your comments were addressed here - would you be willing to approve?

**Review by carlescufi — APPROVED** [review]
Overall structure looks good to me.

**Pull request merged**
